clarity/contracts/arkadiko-oracle-v2-4.clar

```diff
@@ -0,0 +1,252 @@
+;; @contract Arkadiko multisig oracle
+;; @version 2.3
+
+(impl-trait .arkadiko-oracle-trait-v1.oracle-trait)
+
+;; ---------------------------------------------------------
+;; Constants
+;; ---------------------------------------------------------
+
+(define-constant ERR-NOT-AUTHORIZED u8501)
+(define-constant ERR-PRICE-NOT-FOUND u8502)
+
+;; ---------------------------------------------------------
+;; Variables
+;; ---------------------------------------------------------
+
+(define-data-var manual-manager principal (contract-call? .arkadiko-dao get-dao-owner))
+
+;; ---------------------------------------------------------
+;; Maps
+;; ---------------------------------------------------------
+
+(define-map prices-pyth
+  { token: (string-ascii 12) }
+  {
+    decimals: uint,
+    feed-id: (buff 32),
+  }
+)
+
+(define-map prices-dia
+  { token: (string-ascii 12) }
+  {
+    decimals: uint,
+    key: (string-ascii 32),
+  }
+)
+
+(define-map prices-custom
+  { token: (string-ascii 12) }
+  {
+    decimals: uint,
+    key: (string-ascii 32),
+  }
+)
+
+(define-map prices-manual
+  { token: (string-ascii 12) }
+  {
+    decimals: uint,
+    price: uint,
+  }
+)
+
+;; ---------------------------------------------------------
+;; Trait Implementation
+;; ---------------------------------------------------------
+
+(define-public (get-price (token (string-ascii 12)))
+  (fetch-price token)
+)
+
+(define-public (fetch-price (token (string-ascii 12)))
+  (let (
+    (pyth-result (get-pyth-info token))
+  )
+    (if (is-ok pyth-result)
+      pyth-result
+      (let (
+        (dia-result (get-dia-info token))
+      )
+        (if (is-ok dia-result)
+          dia-result
+          (let (
+            (custom-result (get-custom-info token))
+          )
+            (if (is-ok custom-result)
+              custom-result
+              (get-manual-info token)
+            )
+          )
+        )
+      )
+    )
+  )
+)
+
+;; ---------------------------------------------------------
+;; Price Helpers
+;; ---------------------------------------------------------
+
+(define-public (get-pyth-info (token (string-ascii 12)))
+  (let (
+    (price-info (map-get? prices-pyth { token: token }))
+  )
+    (if (is-some price-info)
+      (let (
+        (price (try! (get-price-pyth (get feed-id (unwrap-panic price-info)))))
```

Comment:
Q: since on line 67, we check `(is-ok pyth-result)`, should we do `unwrap-panic` here? I don't know if this can ever panic, but it would make the `fetch-price` method fail? same for logic in the DIA method and custom method

Reply:
A: We could do `unwrap-panic` here as well, but using `try!` makes sure the actual error is returned making it easier to debug in case something goes wrong. In both cases the `is-ok` on line 67 would fail.

Reply:
Ok, perfect!